### ISSUE TITLE
Enrich policies gets into deadlock state if executed before creation

### ISSUE THREAD

**Issue by LucaWintergerst** [elastic/elasticsearch#81248]
**Elasticsearch version** (`bin/elasticsearch --version`): 7.15.1

**Plugins installed**: [] Cloud Defaults

**JVM version** (`java -version`): Cloud

**OS version** (`uname -a` if on a Unix-like system): Cloud

**Description of the problem including expected versus actual behavior**:

Enrich policies gets into deadlock state if executed before creation. 

**Steps to reproduce**:

```

# this one runs forever until it times out in Kibana
PUT /_enrich/policy/my-policy2/_execute
# nothing here, that's expected since the policy does not exist
GET _enrich/policy/my-policy2

# we need an index so the enrich policy can be added later
PUT index2
{
  "mappings": {
    "properties": {
      "area": {
        "type": "long"
      },
      "diff_area": {
        "type": "long"
      },
      "diff_geometry": {
        "type": "geo_shape"
      },
      "diff_id": {
        "type": "text"
      },
      "diff_timestamp": {
        "type": "date"
      },
      "geometry": {
        "type": "geo_shape"
      },
      "id": {
        "type": "text"
      },
      "timestamp": {
        "type": "date"
      }
    }
  }
}

# create policy under same name
PUT _enrich/policy/my-policy2
{
  "geo_match": {
    "indices": "index2",
    "match_field": "diff_geometry",
    "enrich_fields": [
      "id"
    ]
  }
}

DELETE _enrich/policy/my-policy2


```

The delete request then fails with: 

```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "es_rejected_execution_exception",
        "reason" : "Could not obtain lock because policy execution for [my-policy2] is already in progress."
      }
    ],
    "type" : "es_rejected_execution_exception",
    "reason" : "Could not obtain lock because policy execution for [my-policy2] is already in progress."
  },
  "status" : 429
}
```

Only a node restart seems to fix it.

**Provide logs (if relevant)**:

**Comment by martijnvg** [comment]
Thanks @LucaWintergerst for reporting this bug. Looks like we have fixed this already (via #80728), but didn't backport this to the 7.15 branch. On 7.16, when executing the first policy execute api call then this fails with a 404 that the policy doesn't exist. Would you be able to confirm this as well?

**Comment by LucaWintergerst** [comment]
confirmed that this is fixed in 7.16. Apologies, I did not find that PR when searching earlier. I was either looking at issues only or maybe it was because your PR had the NPE but not the API responses that I got. Thanks, glad that this is fixed already!

**Comment by martijnvg** [comment]
Thanks for checking @LucaWintergerst! The PR is more focussed on the cause of why the execute policy api wasn't executing. The side effect of that was that a policy executing would hang, which meant no more policy executions of the same if could happen, which is worse than the NPE.
